File: platforms/android/java/app/msime/client/candidate/CandidatePreeditStylePolicy.java
package app.msime.client;

/**
 * Whether the candidate strip shows what is being spelled.
 *
 * <p>The shared setting is 「候选栏预编辑」 on a touch host, with two values: `pinyin` draws the
 * segmented reading above the candidates, `empty` leaves that line without it. On a phone the strip
 * is the only place a composition is visible at all, so this is a real choice rather than a
 * cosmetic one — some users want the row back for candidates.
 *
 * <p>Two things that share that line are deliberately **not** governed by it.
 *
 * <p>The already-chosen part of a phrase stays visible. The runtime keeps it out of the document at
 * this host's request (`phrase_preedit`) precisely so the user can keep spelling the rest; hiding it
 * here would leave characters the user has already picked neither in the document nor on screen,
 * which is the state `scripts/test-phrase-preedit-hosts.py` exists to prevent.
 *
 * <p>A local input mode's name — 表情, 颜文字, 日期时间 and the rest — stays too. It is a label
 * saying which mode is running, not the text being composed, and removing it would leave the user
 * in a mode with nothing on screen to say so. What the mode is composing beyond its trigger is
 * preedit like any other, and does follow the setting.
 */
public final class CandidatePreeditStylePolicy {
    public static final String PINYIN = "pinyin";
    public static final String EMPTY = "empty";

    private CandidatePreeditStylePolicy() {}

    /** The stored value, with anything unrecognised reading as the shared default. */
    public static String style(String value) {
        return EMPTY.equals(value) ? EMPTY : PINYIN;
    }

    /** Whether the strip draws the composition at all. */
    public static boolean showsComposition(String style) {
        return !EMPTY.equals(style);
    }

    /**
     * The composed text to put on the strip: what was being spelled, or nothing.
     *
     * @param modeName true when the text is a local mode's own name rather than composed input
     */
    public static String composedText(String style, String text, boolean modeName) {
        if (text == null) return "";
        if (modeName || showsComposition(style)) return text;
        return "";
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/candidate/CandidatePreeditStylePolicy.ts
import { LocalInputMode, LocalInputModeDefinition } from "../input/LocalInputMode";

/**
 * What the phone's composition line shows under 「候选栏预编辑」, ported from platforms/android/java/app/msime/client/CandidatePreeditStylePolicy.java.
 *
 * `pinyin` shows the spelling; `empty` leaves the line without it, which is the Windows `candidate_window_preedit_style = "empty"` (`preeditVisible=false`). A 2in1 candidate window takes the whole row away, because it only exists while something is composed. The phone line stays reserved and names the keyboard when idle, so there only the spelling goes and the line keeps its height.
 *
 * Two things that share the line are deliberately not governed by the setting. The already-chosen part of a phrase stays: the runtime keeps it out of the document (`phrase_preedit`) so the user can keep spelling the rest, and hiding it here too would leave characters the user has picked visible nowhere. A local mode's trigger stays as well, because it names the running mode rather than being composed input; what the mode spells beyond it is hidden like any other spelling.
 */
export interface VisibleComposition {
  readonly text: string;
  /** Whether the segment caret is drawn: a hidden spelling has no caret to edit. */
  readonly caret: boolean;
}

export class CandidatePreeditStylePolicy {
  static visible(
    showPreedit: boolean,
    phrasePrefix: string,
    editing: string,
    localMode: string,
  ): VisibleComposition {
    const prefix: string = phrasePrefix ?? "";
    const text: string = editing ?? "";
    if (showPreedit) {
      return { text: text, caret: true };
    }
    const spelling: string = text.startsWith(prefix) ? text.substring(prefix.length) : text;
    const mode: LocalInputModeDefinition | null = LocalInputMode.fromPreferenceKey(localMode);
    if (mode !== null && spelling === mode.trigger) {
      return { text: prefix + spelling, caret: false };
    }
    return { text: prefix, caret: false };
  }
}
